src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/EnumerableExtensions.cs

```diff
@@ -59,4 +69,453 @@ public static bool TryGetCount<T>(this IEnumerable sequence, out int count)
         count = 0;
         return false;
     }
+
+    /// <summary>
+    ///  Copies the contents of the sequence to a destination <see cref="Span{T}"/>.
+    /// </summary>
+    /// <typeparam name="T">The type of elements in the sequence.</typeparam>
+    /// <param name="sequence">The sequence to copy items from.</param>
+    /// <param name="destination">The span to copy items into.</param>
+    /// <exception cref="ArgumentException">
+    ///  The destination span is shorter than the source sequence.
+    /// </exception>
+    public static void CopyTo<T>(this IEnumerable<T> sequence, Span<T> destination)
+    {
+        // First, check a few common cases.
+        switch (sequence)
+        {
+            // We specifically test ImmutableArray<T> to avoid boxing it as an IReadOnlyList.
```

Comment:
It will certainly be faster to use the `ImmutableArray<T>` version, but isn't it already boxed?

Reply:
Oh, yes, of course. I'll fix that comment. Thanks!

Reply:
Since it's already boxed, I ended up removing this case and letting it fall through to `IReadOnlyList<T>`.